### PR TITLE
Add filter validation to query*Features

### LINKED ISSUE #2349: Validate filters passed to queryRenderedFeatures
**mapbox-gl-js version**: v0.16.0

### Steps to Trigger Behavior
queryRenderedFeatures filter is not working as expected as it is returning more features than desired.

1.  See fiddle: https://jsfiddle.net/rraponi/3pznu2ym/

### Expected Behavior
The cursor should change only over the markers that are displayed via
the filter that should only return features with the property of "test" that equal "bar".

 map.on('mousemove', function(e) {
    var features = map.queryRenderedFeatures(e.point, {
      filter: '["==", "test", "bar"]'
    });
    map.getCanvas().style.cursor = features.length ? 'default' : '';
  });

### Actual Behavior
More features are being returned and the cursor change is triggered in unexpected places.

### PR DESCRIPTION
**Pull request by lucaswoj**
fixes #2349

cc @ansis

### CODE REVIEW

File: js/style/style.js [hunk 1]
```diff
@@ -633,6 +633,15 @@ Style.prototype = util.inherit(Evented, {
     },
 
     queryRenderedFeatures: function(queryGeometry, params, classes, zoom, bearing) {
+        if (params.filter) {
+            if (validateStyle.emitErrors(this, validateStyle.filter({
+                key: 'queryRenderedFeatures.filter',
+                value: params.filter,
+                style: this.serialize(),
+                styleSpec: styleSpec
+            }))) return this;
```
Comment (ansis) [inline js/style/style.js]: This is a different return type than is expected. Should this throw instead?

File: js/style/style.js [hunk 2]
```diff
@@ -643,6 +652,22 @@ Style.prototype = util.inherit(Evented, {
...
     },
 
+    querySourceFeatures: function(sourceID, params) {
+        if (params.filter) {
+            if (validateStyle.emitErrors(this, validateStyle.filter({
+                key: 'querySourceFeatures.filter',
+                value: params.filter,
+                style: this.serialize(),
+                styleSpec: styleSpec
+            }))) return this;
```
Comment (ansis) [inline js/style/style.js]: same question as https://github.com/mapbox/mapbox-gl-js/pull/2379/files#r58271553

File: js/style/style.js (hunk 1 above)
Comment (lucaswoj) [inline js/style/style.js]: Yes, the return value should be an empty array.

We don't throw anywhere else in `Style`, even when we can. I think there's value in consistency in error reporting.

Comment (lucaswoj) [inline js/style/style.js]: I take that back. There are some places we do throw in `Style`. This should be a throw.

**Comment by ansis** [comment]
looks good to me!

**Pull request merged**